[core-auth] Create @azure/core-auth for base authentication interfaces

daviwil opened the pull request:
This PR adds a new `@azure/core-auth` library which contains the base authentication interfaces used in `@azure/core-http` and `@azure/identity`.

Fixes #3958.

**Questions for Reviewers**

- How do we like `core-auth` as a name?  Should it be `core-authentication`?
- Is there a better way to get access to `AbortSignalLike` that doesn't require copying the source into `core-auth`?  Pulling in `abort-controller` as a dependency seems like the wrong thing to do.

**Tasks**
- [x] Update `@azure/core-http` and `@azure/core-amqp` to depend on this library
- [x] Re-export these types from `@azure/core-http`

bterlson commented on sdk/core/core-auth/review/core-auth.api.md:
````diff
@@ -0,0 +1,39 @@
+```ts
+
+// @public
+export interface AbortSignalLike {
````
Let's import this from core/abort-controller.

daviwil replied:
Done!

daviwil commented on sdk/core/core-amqp/src/cbs.ts:
```diff
@@ -129,15 +129,15 @@ export class CbsClient {
         );
         log.cbs(
           "[%s] Successfully created the cbs sender '%s' and receiver '%s' " +
-            "links over cbs session.",
```
Looks like Emacs decided to eat some indentation, I'll revert the whitespace changes before commit

daviwil commented:
Alright, I've got this rebased on `master` and updated to use `@azure/abort-controller`.  I've also updated the `core-auth` package version to `1.0.0-preview.2` since that's the first Azure.Core version where it will appear, does that make sense @bterlson?

daviwil commented:
Confirmed with @bterlson and @ramya-rao-a offline that `1.0.0-preview.1` is the appropriate version.  All updates are in, merging this now!
